web/controllers/api/key_controller.ex

```diff
@@ -45,18 +45,66 @@ defmodule HexWeb.API.KeyController do
   end
 
   def delete(conn, %{"name" => name}) do
-    if key = HexWeb.Repo.one(Key.get(name, conn.assigns.user)) do
-      {:ok, _} =
-        Ecto.Multi.new
-        |> Ecto.Multi.delete(:key, key)
-        |> audit(conn, "key.remove", key)
-        |> HexWeb.Repo.transaction
+    user = conn.assigns.user
 
-      conn
-      |> api_cache(:private)
-      |> send_resp(204, "")
+    if key = HexWeb.Repo.one(Key.get(name, user)) do
+      Ecto.Multi.new
+      |> Ecto.Multi.update_all(:key, Key.revoke(user, name), [])
+      |> audit(conn, "key.remove", key)
+      |> HexWeb.Repo.transaction
+      |> case do
+        {:ok, _} ->
+          if key.id === user.current_key_id do
+            key = HexWeb.Repo.get!(assoc(user, :keys), key.id)
```

Comment:
See my note below about the changeset. By using a changeset we don't have to re-fetch the key, `Multi.update` will store the updated key on the `Multi` struct.